Sherlock/replicator.h

```diff
@@ -0,0 +1,285 @@
+            sherlock::constants::kDefaultNamespaceName) {
+    CheckRemoteSchema();
+  }
+
+  explicit SubscribableRemoteStream(const std::string& remote_stream_url,
+                                    const std::string& top_level_name,
+                                    const std::string& namespace_name)
+      : stream_(remote_stream_url),
+        schema_(
+            Value<reflection::ReflectedTypeBase>(reflection::Reflector().ReflectType<T_STREAM_ENTRY>()).type_id,
+            top_level_name,
+            namespace_name) {
+    CheckRemoteSchema();
+  }
+  ~SubscribableRemoteStream() {}
+
+  template <typename F>
+  RemoteSubscriberScope<F, T_STREAM_ENTRY> Subscribe(F& subscriber,
+                                                     uint64_t start_idx = 0u,
+                                                     std::function<void()> done_callback = nullptr) {
+    static_assert(current::ss::IsStreamSubscriber<F, T_STREAM_ENTRY>::value, "");
+    try {
+      return RemoteSubscriberScope<F, T_STREAM_ENTRY>(stream_, subscriber, start_idx, done_callback);
+    } catch (const current::sync::InDestructingModeException&) {
+      CURRENT_THROW(StreamInGracefulShutdownException());
+    }
+  }
+
+ private:
+  void CheckRemoteSchema() {
+    const auto response = HTTP(GET(stream_.ObjectAccessorDespitePossiblyDestructing().url_ + "/schema.simple"));
+    if (static_cast<int>(response.code) == 200) {
+      SubscribableSherlockSchema remote_schema =
+          ParseJSON<SubscribableSherlockSchema>(response.body);
+      if (remote_schema != schema_) {
+        CURRENT_THROW(RemoteStreamSchemaInvalidException());
+      }
+    } else {
+      CURRENT_THROW(RemoteStreamDoesNotRespondException());
+    }
+  }
+
+  ScopeOwnedByMe<RemoteStream> stream_;
+  SubscribableSherlockSchema schema_;
+};
+
+template <typename T_STREAM>
+struct StreamReplicatorImpl {
+  using EntryResponse = current::ss::EntryResponse;
+  using TerminationResponse = current::ss::TerminationResponse;
+  using entry_t = typename T_STREAM::entry_t;
+  using publisher_t = typename T_STREAM::publisher_t;
+
+  StreamReplicatorImpl(T_STREAM& stream) : stream_(stream) { stream.MovePublisherTo(*this); }
+  ~StreamReplicatorImpl() { stream_.AcquirePublisher(std::move(publisher_)); }
+
+  void AcceptPublisher(std::unique_ptr<publisher_t> publisher) { publisher_ = std::move(publisher); }
+
+  EntryResponse operator()(entry_t&& entry, idxts_t current, idxts_t) {
+    assert(publisher_);
+    publisher_->Publish(std::move(entry), current.us);
+    return EntryResponse::More;
+  }
+
+  EntryResponse operator()(const entry_t& entry, idxts_t current, idxts_t) {
+    assert(publisher_);
+    publisher_->Publish(entry, current.us);
+    return EntryResponse::More;
+  }
+
+  EntryResponse EntryResponseIfNoMorePassTypeFilter() const { return EntryResponse::More; }
+  TerminationResponse Terminate() const { return TerminationResponse::Terminate; }
+
+ private:
+  T_STREAM& stream_;
```

Comment:
Nit: `using stream_t` above as well, and `stream_t` here.

Also, upon introducing the `type_alias_t` semantics, we've moved from `T_STREAM` to just `STREAM` in template parameter names. Thanks!